Issue #49: Add Torrent interface

nathancyam opened the pull request:
This adds a form wrapper when viewing one's own torrents. It identifies whether the user that you are viewing is you and then wraps the entire table in a form so you can change their properties.

I am still getting a grip on Python and Flask, so feedback would be very welcome.

nyaadev commented on nyaa/routes.py:
```diff
@@ -267,14 +267,21 @@ def view_user(user_name):
         'per_page': per_page
     }
 
+    is_logged_in_user_account = False
```
This can be done in the jinja templates too.

nyaadev commented on nyaa/static/css/main.css:
```diff
@@ -214,3 +214,56 @@ table.torrent-list thead th.sorting_desc:after {
+}
+
+#torrent-mass-action-form > * {
+    margin-bottom: 1em;
```
mixing spaces with tabs

nyaadev commented on nyaa/static/js/main.js:
```diff
@@ -149,6 +149,46 @@ document.addEventListener("DOMContentLoaded", function() {
   });
 });
 
+document.addEventListener('DOMContentLoaded', function () {
```
should be tabs, though someone already mixed in that one before you...

nyaadev commented on nyaa/templates/search_results.html:
```diff
@@ -50,51 +91,58 @@
-				<td class="text-center">{{ torrent.stats.download_count }}</td>
-				{% endif %}
-				{% endif %}
+
```
you changed this from tabs to spaces

nyaadev commented:
A screenshot btw:
![image](https://cloud.githubusercontent.com/assets/28658364/26208865/77097432-3beb-11e7-8acb-e8c8fab62197.png)

Always good to include one.

nathancyam commented:
@nyaadev Sure, I can add the actions for admin users. There are a few things too that are missing now that I looked into it:

- [ ] There ought to be some client side validation that stops users submitting the form when no torrents are selected
- [ ] Fix the design

Also, what browser are you using? I didn't seem to get the disproportionate column widths while I testing. Can you also confirm that we are using tabs everywhere instead of spaces?

nyaadev commented:
We are using 4 spaces in python and tabs (with width 4) in HTML, CSS and JS.
I'm using Chromium.
Mobile optimizations are missing for the rest of the site as well, that can be done later.

nathancyam commented:
![torrent_interface_mobile](https://cloud.githubusercontent.com/assets/2491442/26253004/329febcc-3cf5-11e7-9b21-de79074b9289.png)

Current mockup of the presentation of torrents on mobile widths. Thoughts?

nathancyam commented:
I've rebased my work onto the current upstream master branch and _hopefully_ addressed the issues that you pointed out @nyaadev.

The result is as follows:

![final_table](https://cloud.githubusercontent.com/assets/2491442/26305400/79665c50-3f32-11e7-99cc-ba2a42b00e35.png)

Let me know if Chromium still doesn't want to play nice.

nathancyam commented:
Sorry to bother you @nyaadev, but can you pull this once you get the chance?

Also, nice works on the comments.

aldacron commented:
Sorry, can you resolve the conflicts?

nathancyam commented:
@aldacron Sure thing. Done.

aldacron commented:
ah, fuck @nathancyam can you resolve the conflicts again?

nathancyam commented:
@aldacron Done.

sharkykh commented on nyaa/static/css/main.css:
```diff
@@ -365,3 +366,96 @@ h6:hover .header-anchor {
+}
+
+#torrent-mass-action-form .form-group > * {
+    margin: 0.5em 0;
```
Spaces to tab

sharkykh commented on nyaa/static/css/main.css:
```diff
@@ -365,3 +366,96 @@ h6:hover .header-anchor {
+		flex-direction: row;
+
+		width: 75%;
+        justify-content: flex-end;
```
Spaces to tab

sharkykh commented on nyaa/static/css/main.css:
```diff
@@ -365,3 +366,96 @@ h6:hover .header-anchor {
+	#torrent-mass-action-form .form-group .category-select select {
+        border-radius: 0;
+		border-left: 0;
+    }
```
Spaces to tab

sharkykh commented on nyaa/static/js/main.js:
```diff
@@ -188,7 +188,47 @@ document.addEventListener("DOMContentLoaded", function() {
 	}
 });
 
-//
```
I suggest that you **_do_** add at least a one-liner to describe what the following lines do, like most of the functions above.

sharkykh commented on nyaa/static/js/main.js:
```diff
@@ -188,7 +188,47 @@ document.addEventListener("DOMContentLoaded", function() {
 });
 
-//
+document.addEventListener('DOMContentLoaded', function () {
```
Spaces to tabs in this function, to match the coding style.

sharkykh commented on nyaa/templates/_formhelpers.html:
```diff
@@ -139,3 +139,54 @@
 	</div>
 {% endmacro %}
+
+{% macro render_category_select(name) %}
```
Change spaces to tabs all over this addition,
to match the coding style.

aldacron replied:
this is nice, but my main concern is that in layout.html we have dupe code for setting categories. would it be easy to factor it out so they use the same macro?

nathancyam replied:
I can do that, but `layout.html` adds some conditional logic to determine which category should be selected (i.e. `<option selected />`) based on the search query. This sort of logic isn't really present in the macro, but what I think I can do instead is to enhance it where you can pass an expression that evaluates to a boolean which then determines which option should have the `selected` attribute. Thoughts?

sharkykh commented on nyaa/templates/search_results.html:
```diff
@@ -17,11 +19,122 @@
 
+{% macro render_torrent_card(torrent, is_current_user) %}
+	{# This currently is not in use yet, but it's a card-like styling for torrents on mobile widths #}
+    {% set cat_id = (torrent.main_category_id|string) + '_' + (torrent.sub_category_id|string) if use_elastic else (torrent.main_category.id|string) + '_' + (torrent.sub_category.id|string) %}
```
Change spaces to tabs all over this addition,
to match the coding style.

nathancyam commented:
My thanks to @sharkykh. @aldacron, please merge at your convenience.

aldacron commented on nyaa/forms.py:
```diff
@@ -174,6 +174,9 @@ def validate_category(form, field):
 
 class UploadForm(FlaskForm):
+    class Meta:
+        csrf = False
```
any idea why we're adding this back? afaik, we _do_ want csrf in UploadForm

nathancyam replied:
It's probably something that git or I did by accident. I'll amend it when I can.

aldacron commented on nyaa/templates/search_results.html:
```diff
@@ -17,11 +19,120 @@
 {% endif %}
 {% endif %}
 
+{% macro render_torrent_card(torrent, is_current_user) %}
```
if this isn't used, can we not have unused code to keep the codebase clean?

nathancyam commented:
Hi @aldacron, changes have been made. I've made the category dropdowns reusable as you've suggested.

aldacron commented on nyaa/templates/layout.html:
```diff
@@ -275,11 +231,7 @@
-										{{ cat_name }}
-									</option>
-									{% endfor %}
+									{{ render_category_options(selected_cat_id=search.category, has_selected_option=search is defined) }}
```
I was testing this locally. in the /login page, this crashes because search isn't passed in -- could this be fixed and could you just run through a few scenarios locally to make sure nothing else crashes? thanks!

nathancyam replied:
Oomph. Sure, thanks for pointing this out.

nathancyam commented:
Hi @aldacron, I've updated the relevant section of `layout.html` to check if the search variable is available in the template. Additionally, to make sure that this doesn't happen again, I've added some very basic tests for the following pages, so people can check if their changes:

- Index
- RSS
- Login
- Torrent Upload
- Help
- Rules
- Registration

I don't know how this could tie in with Travis, but I've left some instructions on the readme.

TheAMM commented:
Ehh, I'd keep tests in another PR, especially since these are very barebones. But @aldacron is handling this so it's up to him.

(Also: `progressbar2` is added back to the dependencies (we use `-33`), the test file is using tabs instead of spaces and while I don't know if Flask testapp does something fancy with Flask-SQA, I doubt that `app.config['DATABASE']` will work with our setup.)

sharkykh commented:
@TheAMM I'm already working on the Travis side of things (tests), so you'll have that ready.

sharkykh commented:
@aldacron 
Umm, nothing really.
It'll have one small conflict which is easy to resolve:
I only changed the tabs in `test_nyaa.py` to spaces. The code is exactly the same.

aldacron commented:
something keeps coming up and preventing me from merging... if you visit the homepage, the S/D/L count is missing. perhaps an issue happened during the merge?

aldacron commented:
in addition, another bug I found:

I have an admin account, with only 1 "deleted" torrent, when I click select visible items, it crashes: 

![crash](https://user-images.githubusercontent.com/28661185/26914339-8695723a-4bd5-11e7-85db-93a09c2e9069.jpg)

![crash2](https://user-images.githubusercontent.com/28661185/26914344-8a74c018-4bd5-11e7-8de4-2abd6dd2910f.jpg)

aldacron commented on nyaa/forms.py:
```diff
@@ -286,13 +286,120 @@ def validate_user_class(form, field):
+        self.selected_torrents = torrents
+
+        if self.action.data == 'move_category':
+            [primary, secondary] = self.category.data.split('_')
```
see comment for crash here

nathancyam replied:
Oh, I think I know what happened. Did you try to move the torrent to a new category, but didn't change the default category option?

sharkykh commented:
> something keeps coming up and preventing me from merging... if you visit the homepage, the S/D/L count is missing. perhaps an issue happened during the merge?

(Fix)
```diff
diff --git a/nyaa/templates/search_results.html b/nyaa/templates/search_results.html
index 2398ca3..8da34e4 100644
--- a/nyaa/templates/search_results.html
+++ b/nyaa/templates/search_results.html
@@ -150,6 +150,17 @@
 					{% else %}
 						<td class="text-center" data-timestamp="{{ torrent.created_utc_timestamp | int }}">{{ torrent.created_time.strftime('%Y-%m-%d %H:%M') }}</td>
 					{% endif %}
+					{% if config.ENABLE_SHOW_STATS %}
+					{% if use_elastic %}
+					<td class="text-center" style="color: green;">{{ torrent.seed_count }}</td>
+					<td class="text-center" style="color: red;">{{ torrent.leech_count }}</td>
+					<td class="text-center">{{ torrent.download_count }}</td>
+					{% else %}
+					<td class="text-center" style="color: green;">{{ torrent.stats.seed_count }}</td>
+					<td class="text-center" style="color: red;">{{ torrent.stats.leech_count }}</td>
+					<td class="text-center">{{ torrent.stats.download_count }}</td>
+					{% endif %}
+					{% endif %}
 				</tr>
 			{% endfor %}
 			</tbody>
```

aldacron commented on nyaa/templates/search_results.html:
```diff
@@ -42,80 +90,74 @@
-				<td class="text-center" data-timestamp="{{ torrent.created_utc_timestamp | int }}">{{ torrent.created_time.strftime('%Y-%m-%d %H:%M') }}</td>
-				{% endif %}
-				
-				{% if config.ENABLE_SHOW_STATS %}
```
seems like this part got omitted so stats don't show at all (see screenshot)

nathancyam replied:
This is what I get when I try to merge shit too late at night.

nathancyam commented:
@aldacron hopefully this resolves the issues that you identified.

sharkykh commented on requirements.txt:
```diff
@@ -28,6 +28,7 @@ mysqlclient==1.3.10
 orderedset==2.0
 packaging==16.8
 passlib==1.7.1
+progressbar2==3.20.0
```
You still have progressbar2 here, you need to remove this.

sharkykh commented on tests/test_nyaa.py:
```diff
@@ -50,4 +50,4 @@ def test_registration_url(self):
 
 if __name__ == '__main__':
-    unittest.main()
+    unittest.main()
```
Python files need to have a blank line at the end.

sharkykh commented:
Hmm, feels like this got stuck.

@nathancyam 
- I've made significant changes to the app structure in the latest commits.
- I **squashed and rebased (some) your work you did in this PR** against #254, **_but only locally_**.
  Why #254? Because of the changes to the categories.
- The massive whitespace changes made it quite hard to see what was actually changed, so I ditched them and went for a simpler diff instead.
We can always do this in a separate branch/pull request, one that we **know** only changes whitespace.

I've pushed the squashed and rebased commit [to my fork](https://github.com/sharkykh/nyaa/commits/torrent-interface).

I _won't_ push to this PR until:
1. I've tested that it still works and looks as expected.
2. I've received a go-ahead to merge #254 (ping @aldacron)
3. I've created a db revision (for migration).
4. I've gotten your approval.

nathancyam commented:
@sharkykh You have my explicit approval. I've been quite busy nowadays, so I haven't really been actively rebasing the branch on the current state of the master branch.

sharkykh commented:
I worked on it for a bit yesterday, and it works, but I'm not done with it yet.
Ideally the code for `UserTorrentMassAction` form should be similar to the `Upload` form.

I also have a different ideas for the design:
## 1.
![image](https://user-images.githubusercontent.com/10238474/28746908-64affb14-749c-11e7-9572-4f7161c2736b.png)

## 2.
![image](https://user-images.githubusercontent.com/10238474/28746877-a0a467e6-749b-11e7-9055-15845dfb25cb.png)

Nalapl3 commented:
Is this pull request going to be merged eventually? It's been quite a time since last activity here.

Arylide commented:
Seemingly abandoned PR, closing for now. If anyone wants to work on this please submit a new PR and reference this one. Also depends on #254 which is also potentially abandoned.
